src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -0,0 +1,449 @@
+			*strrchr (assembly_name, '.') = '\0';
+
+			md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+			mono_register_config_for_assembly (assembly_name, (const char*)map_info.area);
+
+			continue;
+		}
+
+		if (!(utils.ends_with (file_name, ".dll") || utils.ends_with (file_name, ".exe")))
+			continue;
+
+		if (entry_is_overridden)
+			continue;
+
+		size_t alloc_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, sizeof(void*), bundled_assemblies_count + 1);
+		bundled_assemblies = reinterpret_cast<MonoBundledAssembly**> (utils.xrealloc (bundled_assemblies, alloc_size));
+		MonoBundledAssembly *cur = bundled_assemblies [bundled_assemblies_count] = reinterpret_cast<MonoBundledAssembly*> (utils.xcalloc (1, sizeof (MonoBundledAssembly)));
+		++bundled_assemblies_count;
+
+		md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+		cur->name = utils.monodroid_strdup_printf ("%s", strstr (file_name, prefix) + prefix_len);
+		log_warn (LOG_DEFAULT, "cur->name == %s", cur->name);
+		cur->data = (const unsigned char*)map_info.area;
+
+		// MonoBundledAssembly::size is const?!
+		unsigned int *psize = (unsigned int*) &cur->size;
+		*psize = static_cast<unsigned int>(file_size);
+
+		if (utils.should_log (LOG_ASSEMBLY)) {
+			const char *p = (const char*) cur->data;
+
+			char header[9];
+			for (size_t i = 0; i < sizeof(header)-1; ++i)
+				header[i] = isprint (p [i]) ? p [i] : '.';
+			header [sizeof(header)-1] = '\0';
+
+			log_info_nocheck (LOG_ASSEMBLY, "file-offset: % 8x  start: %08p  end: %08p  len: % 12i  zip-entry:  %s name: %s [%s]",
+			                  (int) data_offset, cur->data, cur->data + *psize, (int) file_size, file_name, cur->name, header);
+		}
+	}
+}
+
+bool
+EmbeddedAssemblies::zip_read_cd_info (int fd, uint32_t& cd_offset, uint32_t& cd_size, uint16_t& cd_entries)
+{
+	// The simplest case - no file comment
+	off_t ret = ::lseek (fd, -ZIP_EOCD_LEN, SEEK_END);
+	if (ret < 0) {
+		log_error (LOG_DEFAULT, "Unable to seek into the APK to find ECOD: %s (ret: %d; errno: %d)", std::strerror (errno), ret, errno);
+		return false;
+	}
+
+	uint8_t eocd[ZIP_EOCD_LEN];
+	ssize_t nread = ::read (fd, eocd, static_cast<size_t>(ZIP_EOCD_LEN));
+	if (nread < 0 || nread != ZIP_EOCD_LEN) {
+		log_error (LOG_DEFAULT, "Failed to read EOCD from the APK: %s (nread: %d; errno: %d)", std::strerror (errno), nread, errno);
+		return false;
+	}
+
+	size_t index = 0; // signature
+	uint8_t signature[4];
+
+	if (!zip_read_field (eocd, ZIP_EOCD_LEN, index, signature)) {
+		log_error (LOG_DEFAULT, "Failed to read EOCD signature");
+		return false;
+	}
+
+	if (memcmp (signature, ZIP_EOCD_MAGIC, sizeof(signature)) == 0) {
+		return zip_extract_cd_info (eocd, ZIP_EOCD_LEN, cd_offset, cd_size, cd_entries);
+	}
+
+	// Most probably a ZIP with comment
+	size_t alloc_size = 65535 + ZIP_EOCD_LEN; // 64k is the biggest comment size allowed
+	ret = lseek (fd, static_cast<off_t>(-alloc_size), SEEK_END);
```

Comment:
Should this be `::lseek()`, as done elsewhere?  Or should elsewhere use `lseek()`?  If `::` *shouldn't* be used, the `::read()` calls should also be fixed up.